### ISSUE TITLE
Server crashes at startup with Kafka Engine table

### ISSUE THREAD

**Issue by MrSsunlight** [ClickHouse/ClickHouse#26643]
Frequent core dump after startup；
Insert data by using the kafka engine in protobuf format，The same data will not be a problem in version 20.3.11.97，and the system resource usage is very low。
```
2021.07.21 12:53:07.790148 [ 405119 ] {} <Warning> Application: Listen [::]:8123 failed: Poco::Exception. Code: 1000, e.code() = 0, e.displayText() = DNS error: EAI: Address family for hostname not supported (version 21.7.4.18 (official build)). If it is an IPv6 or IPv4 address and your host has disabled IPv6 or IPv4, then consider to specify not disabled IPv4 or IPv6 address to listen in <listen_host> element of configuration file. Example for disabled IPv6: <listen_host>0.0.0.0</listen_host> . Example for disabled IPv4: <listen_host>::</listen_host>
2021.07.21 12:53:07.790550 [ 405119 ] {} <Warning> Application: Listen [::]:9000 failed: Poco::Exception. Code: 1000, e.code() = 0, e.displayText() = DNS error: EAI: Address family for hostname not supported (version 21.7.4.18 (official build)). If it is an IPv6 or IPv4 address and your host has disabled IPv6 or IPv4, then consider to specify not disabled IPv4 or IPv6 address to listen in <listen_host> element of configuration file. Example for disabled IPv6: <listen_host>0.0.0.0</listen_host> . Example for disabled IPv4: <listen_host>::</listen_host>
2021.07.21 12:53:07.790985 [ 405119 ] {} <Warning> Application: Listen [::]:9009 failed: Poco::Exception. Code: 1000, e.code() = 0, e.displayText() = DNS error: EAI: Address family for hostname not supported (version 21.7.4.18 (official build)). If it is an IPv6 or IPv4 address and your host has disabled IPv6 or IPv4, then consider to specify not disabled IPv4 or IPv6 address to listen in <listen_host> element of configuration file. Example for disabled IPv6: <listen_host>0.0.0.0</listen_host> . Example for disabled IPv4: <listen_host>::</listen_host>
2021.07.21 12:53:07.791375 [ 405119 ] {} <Warning> Application: Listen [::]:9004 failed: Poco::Exception. Code: 1000, e.code() = 0, e.displayText() = DNS error: EAI: Address family for hostname not supported (version 21.7.4.18 (official build)). If it is an IPv6 or IPv4 address and your host has disabled IPv6 or IPv4, then consider to specify not disabled IPv4 or IPv6 address to listen in <listen_host> element of configuration file. Example for disabled IPv6: <listen_host>0.0.0.0</listen_host> . Example for disabled IPv4: <listen_host>::</listen_host>
2021.07.21 12:54:59.436768 [ 412973 ] {} <Fatal> BaseDaemon: ########################################
2021.07.21 12:54:59.436858 [ 412973 ] {} <Fatal> BaseDaemon: (version 21.7.4.18 (official build), build id: 697A9902DCE7A291C0E5CA723B17481126142A81) (from thread 405325) (no query) Received signal Segmentation fault (11)
2021.07.21 12:54:59.436900 [ 412973 ] {} <Fatal> BaseDaemon: Address: 0x28 Access: read. Address not mapped to object.
2021.07.21 12:54:59.436948 [ 412973 ] {} <Fatal> BaseDaemon: Stack trace: 0xf55a9b7 0xf5593b8 0xf557dce 0xfedc168 0xf537685 0xf536ef1 0xfaca439 0xf80f47e 0xfac11b3 0xfac070e 0xfac9e8b 0xfaca454 0xf80f47e 0xfac11b3 0xfac070e 0xfac9e8b 0xfaca50c 0xfaca5a9 0xfac1cbf 0xf537963 0x10259739 0x10258318 0xf767740 0xf7697b7 0xf76a534 0x8d6b91f 0x8d6f203 0x7fcfde4c9ea5 0x7fcfddde69fd
2021.07.21 12:54:59.437126 [ 412973 ] {} <Fatal> BaseDaemon: 1. DB::ReverseIndex<unsigned long, DB::ColumnString>::insert(StringRef const&) @ 0xf55a9b7 in /usr/bin/clickhouse
2021.07.21 12:54:59.437168 [ 412973 ] {} <Fatal> BaseDaemon: 2. COW<DB::IColumn>::mutable_ptr<DB::IColumn> DB::ColumnUnique<DB::ColumnString>::uniqueInsertRangeImpl<char8_t>(DB::IColumn const&, unsigned long, unsigned long, unsigned long, DB::ColumnVector<char8_t>::MutablePtr&&, DB::ReverseIndex<unsigned long, DB::ColumnString>*, unsigned long) @ 0xf5593b8 in /usr/bin/clickhouse
2021.07.21 12:54:59.437195 [ 412973 ] {} <Fatal> BaseDaemon: 3. DB::ColumnUnique<DB::ColumnString>::uniqueInsertRangeFrom(DB::IColumn const&, unsigned long, unsigned long) @ 0xf557dce in /usr/bin/clickhouse
2021.07.21 12:54:59.437243 [ 412973 ] {} <Fatal> BaseDaemon: 4. DB::ColumnLowCardinality::insertRangeFrom(DB::IColumn const&, unsigned long, unsigned long) @ 0xfedc168 in /usr/bin/clickhouse
2021.07.21 12:54:59.437270 [ 412973 ] {} <Fatal> BaseDaemon: 5. void DB::SquashingTransform::append<DB::Block const&>(DB::Block const&) @ 0xf537685 in /usr/bin/clickhouse
2021.07.21 12:54:59.437293 [ 412973 ] {} <Fatal> BaseDaemon: 6. DB::Block DB::SquashingTransform::addImpl<DB::Block const&>(DB::Block const&) @ 0xf536ef1 in /usr/bin/clickhouse
2021.07.21 12:54:59.437321 [ 412973 ] {} <Fatal> BaseDaemon: 7. DB::SquashingBlockOutputStream::write(DB::Block const&) @ 0xfaca439 in /usr/bin/clickhouse
2021.07.21 12:54:59.437349 [ 412973 ] {} <Fatal> BaseDaemon: 8. DB::CountingBlockOutputStream::write(DB::Block const&) @ 0xf80f47e in /usr/bin/clickhouse
2021.07.21 12:54:59.437373 [ 412973 ] {} <Fatal> BaseDaemon: 9. DB::PushingToViewsBlockOutputStream::process(DB::Block const&, DB::PushingToViewsBlockOutputStream::ViewInfo&) @ 0xfac11b3 in /usr/bin/clickhouse
2021.07.21 12:54:59.437404 [ 412973 ] {} <Fatal> BaseDaemon: 10. DB::PushingToViewsBlockOutputStream::write(DB::Block const&) @ 0xfac070e in /usr/bin/clickhouse
2021.07.21 12:54:59.437432 [ 412973 ] {} <Fatal> BaseDaemon: 11. DB::AddingDefaultBlockOutputStream::write(DB::Block const&) @ 0xfac9e8b in /usr/bin/clickhouse
2021.07.21 12:54:59.437455 [ 412973 ] {} <Fatal> BaseDaemon: 12. DB::SquashingBlockOutputStream::write(DB::Block const&) @ 0xfaca454 in /usr/bin/clickhouse
2021.07.21 12:54:59.437478 [ 412973 ] {} <Fatal> BaseDaemon: 13. DB::CountingBlockOutputStream::write(DB::Block const&) @ 0xf80f47e in /usr/bin/clickhouse
2021.07.21 12:54:59.437500 [ 412973 ] {} <Fatal> BaseDaemon: 14. DB::PushingToViewsBlockOutputStream::process(DB::Block const&, DB::PushingToViewsBlockOutputStream::ViewInfo&) @ 0xfac11b3 in /usr/bin/clickhouse
2021.07.21 12:54:59.437525 [ 412973 ] {} <Fatal> BaseDaemon: 15. DB::PushingToViewsBlockOutputStream::write(DB::Block const&) @ 0xfac070e in /usr/bin/clickhouse
2021.07.21 12:54:59.437561 [ 412973 ] {} <Fatal> BaseDaemon: 16. DB::AddingDefaultBlockOutputStream::write(DB::Block const&) @ 0xfac9e8b in /usr/bin/clickhouse
2021.07.21 12:54:59.437594 [ 412973 ] {} <Fatal> BaseDaemon: 17. DB::SquashingBlockOutputStream::finalize() @ 0xfaca50c in /usr/bin/clickhouse
2021.07.21 12:54:59.437617 [ 412973 ] {} <Fatal> BaseDaemon: 18. DB::SquashingBlockOutputStream::writeSuffix() @ 0xfaca5a9 in /usr/bin/clickhouse
2021.07.21 12:54:59.437638 [ 412973 ] {} <Fatal> BaseDaemon: 19. DB::PushingToViewsBlockOutputStream::writeSuffix() @ 0xfac1cbf in /usr/bin/clickhouse
2021.07.21 12:54:59.437665 [ 412973 ] {} <Fatal> BaseDaemon: 20. DB::copyData(DB::IBlockInputStream&, DB::IBlockOutputStream&, std::__1::function<void (DB::Block const&)> const&, std::__1::atomic<bool>*) @ 0xf537963 in /usr/bin/clickhouse
2021.07.21 12:54:59.437689 [ 412973 ] {} <Fatal> BaseDaemon: 21. DB::StorageKafka::streamToViews() @ 0x10259739 in /usr/bin/clickhouse
2021.07.21 12:54:59.437710 [ 412973 ] {} <Fatal> BaseDaemon: 22. DB::StorageKafka::threadFunc(unsigned long) @ 0x10258318 in /usr/bin/clickhouse
2021.07.21 12:54:59.437732 [ 412973 ] {} <Fatal> BaseDaemon: 23. DB::BackgroundSchedulePoolTaskInfo::execute() @ 0xf767740 in /usr/bin/clickhouse
2021.07.21 12:54:59.437763 [ 412973 ] {} <Fatal> BaseDaemon: 24. DB::BackgroundSchedulePool::threadFunction() @ 0xf7697b7 in /usr/bin/clickhouse
2021.07.21 12:54:59.437792 [ 412973 ] {} <Fatal> BaseDaemon: 25. ? @ 0xf76a534 in /usr/bin/clickhouse
2021.07.21 12:54:59.437821 [ 412973 ] {} <Fatal> BaseDaemon: 26. ThreadPoolImpl<std::__1::thread>::worker(std::__1::__list_iterator<std::__1::thread, void*>) @ 0x8d6b91f in /usr/bin/clickhouse
2021.07.21 12:54:59.437844 [ 412973 ] {} <Fatal> BaseDaemon: 27. ? @ 0x8d6f203 in /usr/bin/clickhouse
2021.07.21 12:54:59.437884 [ 412973 ] {} <Fatal> BaseDaemon: 28. start_thread @ 0x7ea5 in /usr/lib64/libpthread-2.17.so
2021.07.21 12:54:59.437931 [ 412973 ] {} <Fatal> BaseDaemon: 29. __clone @ 0xfe9fd in /usr/lib64/libc-2.17.so
2021.07.21 12:54:59.622598 [ 412973 ] {} <Fatal> BaseDaemon: Checksum of the binary: 29FC72CB092D5C4B8F07929DE59AD4B8, integrity check passed.
2021.07.21 12:55:20.183798 [ 405118 ] {} <Fatal> Application: Child process was terminated by signal 11.
```

**Comment by filimonov** [comment]
Can you try to create some isolated reproducible example? Single kafka message & Kafka table + MV definition.

**Comment by MrSsunlight** [comment]
@filimonov 
I cut some data structures, but it won’t have any effect；
**test.proto:**
```
syntax = "proto3";
option optimize_for = SPEED;
message Message {
    uint32 tnow = 1;
    string server = 2;
    string clien = 3;
    uint32 sPort = 4;
    uint32 cPort = 5;
    repeated dd r = 6;
    string method = 7;
}

message dd {
    string name = 1;
    uint32 class = 2;
    uint32 type = 3;
    uint64 ttl = 4;
    bytes data = 5;
}
```
kafka engine table:
```
CREATE TABLE IF NOT EXISTS test.test_queue
(
    `tnow` UInt32,
    `server` String,
    `client` String,
    `sPort` UInt16,
    `cPort` UInt16,
    `r.name` Array(String),
    `r.class` Array(UInt16),
    `r.type` Array(UInt16),
    `r.ttl` Array(UInt32),
    `r.data` Array(String),
    `method` String
)
ENGINE = Kafka
SETTINGS kafka_broker_list = '127.0.0.1:9092', kafka_topic_list = 'test', kafka_group_name = 'group_test', kafka_format = 'Protobuf', kafka_schema = 'test.proto:Message', kafka_num_consumers = 4, kafka_skip_broken_messages = 10000;
```
MV :
```
CREATE MATERIALIZED VIEW test.test_consumer TO test.log AS
SELECT
    toDateTime(a.tnow) AS tnow,
    a.server AS server,
    a.client AS client,
    a.sPort AS sPort,
    a.cPort AS cPort,
    a.`r.name` AS `r.name`,
    a.`r.class` AS `r.class`,
    a.`r.type` AS `r.type`,
    a.`r.ttl` AS `r.ttl`,
    a.`r.data` AS `r.data`,
    a.method AS method
FROM test.test_queue AS a;
```
log table:
```
CREATE TABLE test.log
(
    `tnow` DateTime CODEC(DoubleDelta, LZ4),
    `server` LowCardinality(String),
    `client` LowCardinality(String),
    `sPort` LowCardinality(UInt16),
    `cPort` UInt16 CODEC(T64, LZ4),
    `r.name` Array(String),
    `r.class` Array(LowCardinality(UInt16)),
    `r.type` Array(LowCardinality(UInt16)),
    `r.ttl` Array(LowCardinality(UInt32)),
    `r.data` Array(String),
    `method` LowCardinality(String)
)
ENGINE = MergeTree
PARTITION BY toYYYYMMDD(tnow)
ORDER BY (tnow, server)
TTL toDate(tnow) + toIntervalMonth(1)
SETTINGS index_granularity = 16384, merge_with_ttl_timeout = 7200, storage_policy = 'hdd_in_order';
```

**Comment by filimonov** [comment]
I will try to reproduce the failure.

**Comment by MrSsunlight** [comment]
> I will try to reproduce the failure using your schema

Thanks for looking forward to the recurrence of the problem

**Comment by alexsubota** [comment]
Got same error for v.21.3.3.14, v.21.7.7.47
You can reproduce the problem by creating kafka engine table with kafka_format = 'JSONAsString' and send empty messages.
When setting kafka_format = 'TSV' the problem is not reproducible

**Comment by filimonov** [comment]
> Got same error for v.21.3.3.14, v.21.7.7.47
> You can reproduce the problem by creating kafka engine table with kafka_format = 'JSONAsString' and send empty messages.
> When setting kafka_format = 'TSV' the problem is not reproducible

Your crash is the same as https://github.com/ClickHouse/ClickHouse/issues/19255  (we finally get a reproducible case, thanks). See the fix in #27794

The issue reported by @MrSsunlight seems different.

**Comment by filimonov** [comment]
@MrSsunlight I still can not reproduce the issue you facing using provided schema. Please check the test in  #27822

Maybe there is something specific in your data?
